File: noteStudent/public/js/mediator.js
class Mediator {
    EVENTS = {}; // типы событий
    events = {}; // сами события
    TRIGGERS = {}; // типы триггеров
    triggers = {}; // сами триггеры


    constructor(options) {
        options = options instanceof Object ? options : {};
        this.EVENTS = options.EVENTS;
        this.TRIGGERS = options.TRIGGERS;
        Object.keys(this.EVENTS  ).forEach(key => this.events[this.EVENTS[key]] = []);
        Object.keys(this.TRIGGERS).forEach(key => this.triggers[this.TRIGGERS[key]] = () => null);
    }

    // Подписка на событие
    subscribe(name, func) {
        if (this.events[name] && func instanceof Function) {
            this.events[name].push(func);
        }
    }

    // Вызов события
    call(name, data = null) {
        if (this.events[name]) {
            this.events[name].forEach(event => {
                if (event instanceof Function) {
                    event(data);
                }
            });
        }
    }

    // Вызов триггера
    get(name, data = null) {
        if (this.triggers[name] && this.triggers[name] instanceof Function) {
            return this.triggers[name](data);
        }
        return null;
    }

    // Подписка на триггер
    set(name, func) {
        if (name && func instanceof Function) {
            this.triggers[name] = func;
        }
    }
}

File: noteStudent/application/Mediator.py
class Mediator:
    EVENTS = {}  # типы событий
    events = {}  # списки событий
    TRIGGERS = {}  # типы триггеров
    triggers = {}  # список триггеров

    def __init__(self, types):
        self.EVENTS = types['EVENTS']
        self.TRIGGERS = types['TRIGGERS']
        for key in self.EVENTS.keys():
            self.events.update({self.EVENTS[key]: []})
        for key in self.TRIGGERS.keys():
            self.triggers.update({self.TRIGGERS[key]: lambda x=None: None})

    def __del__(self):
        self.events.clear()
        self.triggers.clear()

    # подписка на события
    def subscribe(self, name, func):
        if name and func:
            self.events.get(name).append(func)

    def getEvents(self):
        return self.EVENTS

    # выозов события
    def call(self, name, data=None):
        if name:
            cbs = self.events.get(name)
            if cbs:
                for cb in cbs:
                    cb(data)

    def getTriggers(self):
        return self.TRIGGERS

    # Установить Триггер
    def set(self, name, func):
        if name and func:
            self.triggers[name] = func

    # Вызвать триггер
    def get(self, name, data=None):
        if name:
            return self.triggers.get(name)(data)
